### ISSUE TITLE
Fails while tagging new release if remote was updated during the release

### ISSUE THREAD

**Issue by DanielHabenicht** [semantic-release/semantic-release#993]
## Current behavior

1. A build with semantic release is triggered. 
2. Semantic release analyzes the commits and determines a version number to tag
3. The remote branch gets updated, while semantic-release is still running (maybe a longer `@semantic-release/exec` command)
3. All commands finish and semantic-release tries to push the new tag to the remote but fails: 
```
Error: Command failed: git push --tags https://remote.gitrepo.com/path/to/project HEAD:master
To https://remote.gitrepo.com/path/to/project
  * [new tag]         v1.1.8 -> v1.1.8
  ! [rejected]        HEAD -> master (fetch first)
 error: failed to push some refs to 'https://remote.gitrepo.com/path/to/project'
 hint: Updates were rejected because the remote contains work that you do
 hint: not have locally. This is usually caused by another repository pushing
 hint: to the same ref. You may want to first integrate the remote changes
 hint: (e.g., 'git pull ...') before pushing again.
 hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

## Expected behavior

Semantic Release pushes the Tag and finishes the build succesfully, even if the remote is updated in the meanwhile. 

## Workaround

Do not push to master while your project is building. 

## Environment

- **semantic-release** version: 15.10.3
- CI environment: Team Foundation Service (Azure DevOps OnPrem)
- Plugins used: `@semantic-release/changelog`, `@semantic-release/exec`, `@iteratec/semantic-release-docker`
- **semantic-release** configuration: 
```
"release": {
    "branch": "master",
    "verifyConditions": [],
    "verifyConfig": [
      "@iteratec/semantic-release-docker"
    ],
    "prepare": [
      {
        "path": "@semantic-release/changelog",
        "changelogFile": "dist/changelog.md"
      },
      {
        "path": "@semantic-release/exec",
        "cmd": "node version.js ${nextRelease.version} ${commits[0].commit.short} ${commits[0].commit.long}"
      },
      {
        "path": "@semantic-release/exec",
        "cmd": "docker build . -f Dockerfile-dist -t dist_project"
      },
      {
        "path": "@semantic-release/exec",
        "cmd": "docker build ./infrastructure/nginx -t project"
      },
      {
        "path": "@iteratec/semantic-release-docker",
        "additionalTags": [
          "latest"
        ],
        "imageName": "project",
        "registryUrl": "remote.registry.com",
        "respositoryName": "project"
      }
    ],
    "publish": [
      {
        "path": "@iteratec/semantic-release-docker"
      }
    ],
    "fail": [],
    "success": []
  }
```
- CI logs: 
```
2018-11-20T14:01:33.8799780Z [14:01:33] [semantic-release] › ✖  An error occurred while running semantic-release: { Error: Command failed: git push --tags https://remote.gitrepo.com/path/to/project HEAD:master
2018-11-20T14:01:33.8823860Z To https://remote.gitrepo.com/path/to/project
2018-11-20T14:01:33.8844900Z  * [new tag]         v1.1.8 -> v1.1.8
2018-11-20T14:01:33.8865570Z  ! [rejected]        HEAD -> master (fetch first)
2018-11-20T14:01:33.8886290Z error: failed to push some refs to 'https://remote.gitrepo.com/path/to/project'
2018-11-20T14:01:33.8905530Z hint: Updates were rejected because the remote contains work that you do
2018-11-20T14:01:33.8926550Z hint: not have locally. This is usually caused by another repository pushing
2018-11-20T14:01:33.8944380Z hint: to the same ref. You may want to first integrate the remote changes
2018-11-20T14:01:33.8964060Z hint: (e.g., 'git pull ...') before pushing again.
2018-11-20T14:01:33.8984530Z hint: See the 'Note about fast-forwards' in 'git push --help' for details.
2018-11-20T14:01:33.8995090Z 
2018-11-20T14:01:33.9003430Z 
2018-11-20T14:01:33.9043480Z     at makeError (/vsts/agent/_work/4/s/node_modules/execa/index.js:174:9)
2018-11-20T14:01:33.9061320Z     at Promise.all.then.arr (/vsts/agent/_work/4/s/node_modules/execa/index.js:278:16)
2018-11-20T14:01:33.9078110Z     at <anonymous>
2018-11-20T14:01:33.9094970Z     at process._tickCallback (internal/process/next_tick.js:188:7)
2018-11-20T14:01:33.9111570Z   code: 1,
2018-11-20T14:01:33.9144140Z   stdout: '',
2018-11-20T14:01:33.9170470Z   stderr: 'To https://remote.gitrepo.com/path/to/project\n * [new tag]         v1.1.8 -> v1.1.8\n ! [rejected]        HEAD -> master (fetch first)\nerror: failed to push some refs to \'https://remote.gitrepo.com/path/to/project\'\nhint: Updates were rejected because the remote contains work that you do\nhint: not have locally. This is usually caused by another repository pushing\nhint: to the same ref. You may want to first integrate the remote changes\nhint: (e.g., \'git pull ...\') before pushing again.\nhint: See the \'Note about fast-forwards\' in \'git push --help\' for details.\n',
2018-11-20T14:01:33.9195980Z   failed: true,
2018-11-20T14:01:33.9212340Z   signal: null,
2018-11-20T14:01:33.9231630Z   cmd: 'git push --tags https://remote.gitrepo.com/path/to/project HEAD:master',
2018-11-20T14:01:33.9251050Z   timedOut: false,
2018-11-20T14:01:33.9268150Z   killed: false }
2018-11-20T14:01:33.9288320Z { Error: Command failed: git push --tags https://remote.gitrepo.com/path/to/project HEAD:master
2018-11-20T14:01:33.9309460Z To https://remote.gitrepo.com/path/to/project
2018-11-20T14:01:33.9329840Z  * [new tag]         v1.1.8 -> v1.1.8
2018-11-20T14:01:33.9350000Z  ! [rejected]        HEAD -> master (fetch first)
2018-11-20T14:01:33.9371550Z error: failed to push some refs to 'https://remote.gitrepo.com/path/to/project'
2018-11-20T14:01:33.9390980Z hint: Updates were rejected because the remote contains work that you do
2018-11-20T14:01:33.9407790Z hint: not have locally. This is usually caused by another repository pushing
2018-11-20T14:01:33.9424640Z hint: to the same ref. You may want to first integrate the remote changes
2018-11-20T14:01:33.9443230Z hint: (e.g., 'git pull ...') before pushing again.
2018-11-20T14:01:33.9463680Z hint: See the 'Note about fast-forwards' in 'git push --help' for details.
2018-11-20T14:01:33.9474210Z 
2018-11-20T14:01:33.9482510Z 
2018-11-20T14:01:33.9499090Z     at makeError (/vsts/agent/_work/4/s/node_modules/execa/index.js:174:9)
2018-11-20T14:01:33.9515870Z     at Promise.all.then.arr (/vsts/agent/_work/4/s/node_modules/execa/index.js:278:16)
2018-11-20T14:01:33.9532630Z     at <anonymous>
2018-11-20T14:01:33.9550010Z     at process._tickCallback (internal/process/next_tick.js:188:7)
2018-11-20T14:01:33.9568250Z   code: [33m1[39m,
2018-11-20T14:01:33.9589140Z   stdout: [32m''[39m,
2018-11-20T14:01:33.9615600Z   stderr: [32m'To https://remote.gitrepo.com/path/to/project\n * [new tag]         v1.1.8 -> v1.1.8\n ! [rejected]        HEAD -> master (fetch first)\nerror: failed to push some refs to \'https://remote.gitrepo.com/path/to/project\'\nhint: Updates were rejected because the remote contains work that you do\nhint: not have locally. This is usually caused by another repository pushing\nhint: to the same ref. You may want to first integrate the remote changes\nhint: (e.g., \'git pull ...\') before pushing again.\nhint: See the \'Note about fast-forwards\' in \'git push --help\' for details.\n'[39m,
2018-11-20T14:01:33.9641080Z   failed: [33mtrue[39m,
2018-11-20T14:01:33.9660980Z   signal: [1mnull[22m,
2018-11-20T14:01:33.9682130Z   cmd: [32m'git push --tags https://remote.gitrepo.com/path/to/project HEAD:master'[39m,
2018-11-20T14:01:33.9702800Z   timedOut: [33mfalse[39m,
2018-11-20T14:01:33.9763440Z   killed: [33mfalse[39m }npm ERR! code ELIFECYCLE
2018-11-20T14:01:33.9781990Z npm ERR! errno 1
2018-11-20T14:01:33.9800530Z npm ERR! project@0.4.0 release: `semantic-release "--no-ci"`
2018-11-20T14:01:33.9818900Z npm ERR! Exit status 1
2018-11-20T14:01:33.9835210Z npm ERR! 
2018-11-20T14:01:33.9852180Z npm ERR! Failed at the project@0.4.0 release script.
2018-11-20T14:01:33.9869170Z npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
2018-11-20T14:01:33.9877910Z 
2018-11-20T14:01:33.9894410Z npm ERR! A complete log of this run can be found in:
2018-11-20T14:01:33.9913050Z npm ERR!     /root/.npm/_logs/2018-11-20T14_01_33_892Z-debug.log
2018-11-20T14:01:34.0111640Z ##[error]npm failed with error: /usr/local/bin/npm failed with return code: 1
```

**Comment by DanielHabenicht** [comment]
It has to do with this line, explitcitly referencing the branch: 
https://github.com/semantic-release/semantic-release/blob/f2ede5bdded568ee1e2d52af608f3c92123030dc/test/helpers/git-utils.js#L226

Git pushes the tag but errors as the branch is behind origin.

**Comment by pvdlg** [comment]
Is your `version.js` or `@iteratec/semantic-release-docker` commit something to the remote Git repo?
Was there something committing to the remote repo while semantic-release was running?

semantic-release verify that the local branch is up to date with the remote at the beginning of the process and will exit if it's not the case. If you have something updating the remote in between this check and the tag creation there is no much we can do.

I guess we could check if the remote is up to date when the push fails so we can print a slightly more useful message, but it's a case that is so rare that I'm not sure it worth it.

We choose to skip the release when the remote is ahead because:
- we don't want to inadvertently erase a commit you might have done in your CI before running semantic-release
- there is no guarantee that doing a pull would work because the local and remote might have a merge conflict (e.g. if you rebased your head commit on the remote)
- You might have run build scripts before running semantic-release, and those would have used an outdated version of the code, therefore the published code would be wrong
- It can creates several unexpected issues because your CI would have started with a certain head and we would modify it (e.g. the CI dashboard will display a head commit that is not the one actually used)

In any case it doesn't really matter that we skip the release on an outdated local branch, because if the remote has been updated it would trigger another build and semantic-release would run without errors for this one.

**Comment by DanielHabenicht** [comment]
No, I just stumbled upon this issue as we were working normally and another PR was merged into master while the release for the previous merged PR was running. Therefore we (the Team Foundation server) committed to the remote while semantic release was running.

The first problem now is, that there was no version published for this commit.

The second problem is that semantic release did not finish in a pristine state as the tag was pushed to the remote and then exited with failure.

I would expect semantic release to run without error, once it verified in the beginning, if the branch is up to date.

**Comment by pvdlg** [comment]
> The first problem now is, that there was no version published for this commit.

How would you push a tag from an outdated local repo? How could you handle any type of discrepancy between local and remote (including new commit pushed, remote branch reset to an older commit, rebase of the head commit on remote) when pushing that tag?

 > The second problem is that semantic release did not finish in a pristine state as the tag was pushed to the remote and then exited with failure.

According to your logs the tag was **not** pushed to the remote. See "error: failed to push some refs to 'https://remote.gitrepo.com/path/to/project'"

Also semantic-release doesn't have control on external events. We can't prevent external things to happen between when we make the check and when we push.

> I would expect semantic release to run without error, once it verified in the beginning, if the branch is up to date.

Again there is no way to achieve that, as there is no way to prevent things to happen outside of semantic-release. The verification happens at a point in time, and the push at another point in time and we can't prevent things to happen in the outside word between two point in time. As far as I know there is no way to put a lock on the remote between those two point in time, so there will always be a race condition there.

**Comment by DanielHabenicht** [comment]
> How would you push a tag from an outdated local repo? How could you handle any type of discrepancy between local and remote (including new commit pushed, remote branch reset to an older commit, rebase of the head commit on remote) when pushing that tag?

Well, in this case, there is no discrepancy between local and remote apart from the remote being one commit ahead, so it shouldn't be a problem to tag the previous commit. 

I do understand that this is a problem in the situations you described. 
However in my way of thinking I am triggering the semantic release in order to create a release for a specific commit. If the commit is outdated before even starting the release it is fine to skip it and redo it at a later time, just as you stated. 
But as the release is started semantic release should go on with every other task following, including tagging the commit, because it just executed all the commands for this version and e.g. already pushed the docker image or released the npm package. 

Maybe this could be hidden behind a `force` option if it is not favourable by the majority, as there are the drawbacks you just described. 

>According to your logs the tag was not pushed to the remote. See "error: failed to push some refs to 'https://remote.gitrepo.com/path/to/project'"

Unfortunately, it was. I can see it in the TFS and the gitlog is also saying it: 
```
An error occurred while running semantic-release: { Error: Command failed: git push --tags https://remote.gitrepo.com/path/to/project HEAD:master
2018-11-20T14:01:33.8823860Z To https://remote.gitrepo.com/path/to/project
2018-11-20T14:01:33.8844900Z  * [new tag]         v1.1.8 -> v1.1.8
2018-11-20T14:01:33.9350000Z  ! [rejected]        HEAD -> master (fetch first)
```
It did Error on the second task Git was doing. 

In order to reproduce the error I tried executing the command only using git (always with the same configuration: remote repo beeing one commit ahead): 
`git push --tags` is working fine
`git push <tag name> origin` is working fine
`git push --tags origin HEAD:master` is erroring with the same error message:
```
To https://remote.gitrepo.com/path/to/project
 * [new tag]         test6 -> test6
 ! [rejected]        HEAD -> master (non-fast-forward)
error: failed to push some refs to 'https://remote.gitrepo.com/path/to/project'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
I don't know the specifics of Git but specifying `HEAD:master` seems to trigger some sort of push apart from pushing the tags. 

>Again there is no way to achieve that, as there is no way to prevent things to happen outside of semantic-release. The verification happens at a point in time, and the push at another point in time and we can't prevent things to happen in the outside word between two point in time. As far as I know there is no way to put a lock on the remote between those two point in time, so there will always be a race condition there.

I understand that handling external events should be taken care of by the build scheduler, e.g. by building only once an hour and thus circumvent this issue. 
But my main point is that it is possible to tag the build but semantic release fails nontheless.

**Comment by DanielHabenicht** [comment]
I think the command `git push --tags origin HEAD:master` is wrongly used. 
At the end of the release, only the tag should be pushed to the remote repo. But by specifying `origin HEAD:master` the branch is also pushed, but can't because origin is already one commit ahead. Furthermore why should semantic release push the branch - it shouldn't alter the source code? 

It should be `git push --tags` or `git push <tag name> origin`

**Comment by pvdlg** [comment]
I didn't realized the tag was created but pushing the head failed.
Should be fixed in #994 .

Regarding https://github.com/semantic-release/semantic-release/issues/993#issuecomment-440708053 it's not completely correct. Pushing to Docker should be done in the `publish` step which happens after the tag is pushed. You should not publish things in a step before `publish`.

**Comment by DanielHabenicht** [comment]
> I didn't realized the tag was created but pushing the head failed.
> Should be fixed in #994 .
> 
> Regarding [#993 (comment)](https://github.com/semantic-release/semantic-release/issues/993#issuecomment-440708053) it's not completely correct. Pushing to Docker should be done in the `publish` step which happens after the tag is pushed. You should not publish things in a step before `publish`.

Yep, I wasn't sure in the beginning as well. Thanks for the quick fix and keep up the great work. I really like this tool. 👍 
You are right, I confused it with building the images. The plugin I am using takes care of publishing during the publish step.
